src/tests/suite/daos_dedup.c

```diff
@@ -0,0 +1,361 @@
+	 * being inserted
+	 */
+	const daos_size_t	dedup_size_increase = 256;
+	daos_size_t		after_first_update;
+	daos_size_t		after_second_update;
+	daos_size_t		delta;
+	int			rc;
+
+	setup_context(&ctx, *state, iod_type, csum_type, oc, dedup_type,
+		      dedup_threshold_setting);
+
+	rc = ctx_update(&ctx);
+	assert_success(rc);
+	after_first_update = get_size(&ctx);
+
+	/** update again with same data */
+	rc = ctx_update(&ctx);
+	assert_success(rc);
+
+	/** if threshold is less than data size, dedup should prevent the extra
+	 * update and therefor the data used from the pool is much less.
+	 * Otherwise, the data used from the pool will be larger.
+	 */
+	after_second_update = get_size(&ctx);
+	delta = after_first_update - after_second_update;
+	if (dedup_threshold_setting == THRESHOLD_LESS_THAN_DATA &&
+	    delta > dedup_size_increase)
+		fail_msg("Pool used size increased by %lu, which is larger "
+			 "than expected size increase of less than or equal "
+			 "to %lu", delta, dedup_size_increase);
+	else if (dedup_threshold_setting == THRESHOLD_GREATER_THAN_DATA &&
+	    delta < dedup_size_increase)
+		fail_msg("Pool used size increased by %lu, which is less "
+			 "than expected size increase of greater than or equal "
+			 "to %lu", delta, dedup_size_increase);
+}
+
+static void
+array_csumoff_deduphash(void **state)
+{
+	with_identical_updates(state,
+			       DAOS_IOD_ARRAY,
+			       DAOS_PROP_CO_CSUM_OFF,
+			       OC_SX,
+			       DAOS_PROP_CO_DEDUP_HASH,
+			       THRESHOLD_LESS_THAN_DATA);
+}
+static void
+array_csumoff_dedupmemcmp(void **state)
+{
+	with_identical_updates(state,
+			       DAOS_IOD_ARRAY,
+			       DAOS_PROP_CO_CSUM_OFF,
+			       OC_SX,
+			       DAOS_PROP_CO_DEDUP_MEMCMP,
+			       THRESHOLD_LESS_THAN_DATA);
+}
+
+static void
+array_csumcrc64_deduphash(void **state)
+{
+	with_identical_updates(state,
+			       DAOS_IOD_ARRAY,
+			       DAOS_PROP_CO_CSUM_CRC64,
+			       OC_SX,
+			       DAOS_PROP_CO_DEDUP_HASH,
+			       THRESHOLD_LESS_THAN_DATA);
+}
+
+static void
+array_csumcrc64_dedupmemcmp(void **state)
+{
+	with_identical_updates(state,
+			       DAOS_IOD_ARRAY,
+			       DAOS_PROP_CO_CSUM_CRC64,
+			       OC_SX,
+			       DAOS_PROP_CO_DEDUP_MEMCMP,
+			       THRESHOLD_LESS_THAN_DATA);
+}
+
+static void
+array_above_threshold(void **state)
+{
+	with_identical_updates(state,
+			       DAOS_IOD_ARRAY,
+			       DAOS_PROP_CO_CSUM_CRC64,
+			       OC_SX,
+			       DAOS_PROP_CO_DEDUP_MEMCMP,
+			       THRESHOLD_GREATER_THAN_DATA);
+}
+
+static int
+setup(void **state)
+{
+	return test_setup(state, SETUP_POOL_CONNECT, true, DEFAULT_POOL_SIZE,
+			  NULL);
+}
+
+#define DEDUP_TEST(dsc, test) { dsc, test, NULL, test_case_teardown }
+static const struct CMUnitTest dedup_tests[] = {
+	DEDUP_TEST("DAOS_DEDUP01: With array type, csums disabled, hash dedup",
+		array_csumoff_deduphash),
+	DEDUP_TEST("DAOS_DEDUP02: With array type, csums disabled, hash memcmp",
```

Comment:
(style)  'DEDUP' may be misspelled - perhaps 'DEDUPE'?
(style)  'DEDUP' may be misspelled - perhaps 'DEDUPE'?